### PR TITLE
Pinned messages implementation #13

### PR DESCRIPTION
**Pull request by g-harel**
please add your reviews

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -90,6 +96,12 @@ dependencies {
     compile 'com.melnykov:floatingactionbutton:1.3.0'
     compile 'com.google.zxing:android-integration:3.1.0'
     compile 'com.squareup.dagger:dagger:1.2.2'
+
+    // Stetho core
```
Comment (g-harel) [inline build.gradle]: don't think this comment adds any value

File: build.gradle [hunk 2]
```diff
@@ -90,6 +96,12 @@ dependencies {
...
+    compile 'com.facebook.stetho:stetho:1.5.0'
+    //If you want to add a network helper
```
Comment (g-harel) [inline build.gradle]: same

File: src/org/thoughtcrime/securesms/OnSwipeTouchListener.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
+package org.thoughtcrime.securesms;
+import android.content.Context;
+import android.view.GestureDetector;
+import android.view.GestureDetector.SimpleOnGestureListener;
+import android.view.MotionEvent;
+import android.view.View;
+import android.view.View.OnTouchListener;
+
+/**
+ * Created by shawnelbaz on 2018-02-09.
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/OnSwipeTouchListener.java]: no comments

File: src/org/thoughtcrime/securesms/OnSwipeTouchListener.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+/**
+ * Detects left and right swipes across a view.
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/OnSwipeTouchListener.java]: no comments + redundant comment

File: src/org/thoughtcrime/securesms/PinMessagesLoader.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import android.database.Cursor;
+
+import javax.annotation.Nullable;
+
+import org.thoughtcrime.securesms.database.DatabaseFactory;
+import org.thoughtcrime.securesms.database.MmsSmsDatabase;
+import org.thoughtcrime.securesms.util.AbstractCursorLoader;
+
+/**
+ * Created by BABO99 on 2018-02-05.
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/PinMessagesLoader.java]: no comments

File: src/org/thoughtcrime/securesms/PinnedMessageAdapter.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                PinnedMessagesHandler handler = new PinnedMessagesHandler(context);
+                handler.handleUnpinMessage(record, DatabaseFactory.getSmsDatabase(context));
+
+                ((ViewGroup)v.getParent().getParent().getParent()).removeAllViews();
+            }
+        });
+    }
+
+    private void setMessageView(MessageRecord record, ViewHolder viewHolder) {
+        // To check if the message is incoming
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/PinnedMessageAdapter.java]: no comments + redundant

File: src/org/thoughtcrime/securesms/PinnedMessageFragment.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+import android.view.ViewGroup;
+import android.database.Cursor;
+import android.support.v4.app.LoaderManager;
+
+import org.thoughtcrime.securesms.crypto.MasterSecret;
+import org.thoughtcrime.securesms.util.ViewUtil;
+
+
+/**
+ * A simple {@link Fragment} subclass.
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/PinnedMessageFragment.java]: no comments

this looks autogenerated ?

File: src/org/thoughtcrime/securesms/PinnedMessageFragment.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+ * Activities that contain this fragment must implement the
+ * {@link PinnedMessageFragment.OnFragmentInteractionListener} interface
+ * to handle interaction events.
+ * Use the {@link PinnedMessageFragment#newInstance} factory method to
+ * create an instance of this fragment.
+ */
+public class PinnedMessageFragment extends Fragment implements LoaderManager.LoaderCallbacks<Cursor> {
+    //private PinnedMessageFragment.PinnedMessageFragmentListener listener;
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/PinnedMessageFragment.java]: this whole class has a bunch of noise, we should probably clean it up

File: src/org/thoughtcrime/securesms/PinnedMessagesHandler.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
+package org.thoughtcrime.securesms;
+
+import android.content.Context;
+
+import org.thoughtcrime.securesms.database.DatabaseFactory;
+import org.thoughtcrime.securesms.database.MessagingDatabase;
+import org.thoughtcrime.securesms.database.model.MessageRecord;
+
+public class PinnedMessagesHandler {
+    Context context;
```
Comment (g-harel) [inline src/org/thoughtcrime/securesms/PinnedMessagesHandler.java:10]: does this have an implicit `public` ?

File: test/androidTest/java/org/thoughtcrime/securesms/util/SaveAttachmentTaskTest.java [hunk 10]
```diff
@@ -193,11 +193,11 @@ private void verifyAttachmentSavedCorrectly(String outputFileName,
                                                 TEST_TIMESTAMP,
                                                 outputFileName);
 
-    Pair<Integer, File> result = saveAttachmentTask.doInBackground(attachment);
+    // Pair<Integer, File> result = saveAttachmentTask.doInBackground(attachment);
```
Comment (g-harel) [inline test/androidTest/java/org/thoughtcrime/securesms/util/SaveAttachmentTaskTest.java:196]: this was my change, not sure why the test was breaking, should be looked into

File: test/unitTest/java/org/thoughtcrime/securesms/BaseUnitTest.java [hunk 11]
```diff
@@ -25,9 +25,9 @@
 import static org.mockito.Matchers.anyInt;
 import static org.mockito.Matchers.anyLong;
 import static org.mockito.Matchers.anyString;
-import static org.powermock.api.mockito.PowerMockito.mock;
+import static org.mockito.Mockito.mock;
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/BaseUnitTest.java:28]: why are these being imports added if nothing else in the file changed ?

File: test/unitTest/java/org/thoughtcrime/securesms/PinnedMessagesMocks.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+import org.thoughtcrime.securesms.database.model.MessageRecord;
+import org.thoughtcrime.securesms.database.MmsDatabase;
+import org.thoughtcrime.securesms.database.MmsSmsDatabase;
+import org.thoughtcrime.securesms.database.SmsDatabase;
+
+import static org.mockito.Mockito.when;
+
+
+/**
+ * Created by BABO99 on 2018-02-09.
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/PinnedMessagesMocks.java]: no comments

File: test/unitTest/java/org/thoughtcrime/securesms/PinnedMessagesMocks.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        when(messageRecordMms.getId()).thenReturn(1l);
+    }
+
+    protected void setUpStaticMessagingDatabase() throws Exception {
+        PowerMockito.mockStatic(DatabaseFactory.class);
+        BDDMockito.given(DatabaseFactory.getInstance(context)).willReturn(databaseFactory);
+        BDDMockito.given(DatabaseFactory.getSmsDatabase(context)).willReturn(smsDatabase);
+        BDDMockito.given(DatabaseFactory.getMmsDatabase(context)).willReturn(mmsDatabase);
+
+        when( smsDatabase.pinMessage(1)).thenReturn(true);
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/PinnedMessagesMocks.java]: this is weird formatting (the extra space)

File: test/unitTest/java/org/thoughtcrime/securesms/database/PinnedMessagesTest.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+import org.junit.Test;
+import org.thoughtcrime.securesms.PinnedMessagesMocks;
+
+import static junit.framework.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+
+/**
+ * Created by BABO99 on 2018-02-08.
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/database/PinnedMessagesTest.java]: no comments

File: test/unitTest/java/org/thoughtcrime/securesms/PinnedMessagesMocks.java [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+import static org.mockito.Mockito.when;
+
+
+/**
+ * Created by BABO99 on 2018-02-09.
+ */
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({DatabaseFactory.class})
+public class PinnedMessagesMocks extends BaseUnitTest {
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/PinnedMessagesMocks.java]: is it a common pattern to have mocks be in the superclass ?

File: test/unitTest/java/org/thoughtcrime/securesms/database/PinnedMessagesTest.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    public void setUp() {
+        messagingDatabase = mock(MessagingDatabase.class);
+        mmsSmsDatabase = mock(MmsSmsDatabase.class);
+
+        super.setUpPinMethod();
+        super.setUpUnpinMethod();
+    }
+
+    @Test
+    // This test is to check the behaviour of the method
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/database/PinnedMessagesTest.java]: no comments + I don't think the comment is correct, these are testing the mocks

File: test/unitTest/java/org/thoughtcrime/securesms/database/PinnedMessagesTest.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+        assertEquals(messagingDatabase.pinMessage(2), false);
+        assertEquals(messagingDatabase.pinMessage(3), false);
+
+        verify(messagingDatabase).pinMessage(3);
+        verify(messagingDatabase).pinMessage(3);
+        verify(messagingDatabase).pinMessage(3);
+    }
+
+    @Test
+    // This test is to check the behaviour of the method
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/database/PinnedMessagesTest.java]: same

File: test/unitTest/java/org/thoughtcrime/securesms/pinmessages/PinnedMesssagesHandlerTests.java [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+import org.thoughtcrime.securesms.database.SmsDatabase;
+import org.thoughtcrime.securesms.PinnedMessagesHandler;
+import org.thoughtcrime.securesms.PinnedMessagesMocks;
+
+import static junit.framework.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+/**
+ * Created by BABO99 on 2018-02-09.
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/pinmessages/PinnedMesssagesHandlerTests.java]: no comments

File: test/unitTest/java/org/thoughtcrime/securesms/pinmessages/PinnedMesssagesHandlerTests.java [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+        super.mmsDatabase          = mock(MmsDatabase.class);
+        super.smsDatabase          = mock(SmsDatabase.class);
+
+        super.setUpPinMethod();
+        super.setUpUnpinMethod();
+        super.setUpMessageRecord();
+        super.setUpStaticMessagingDatabase();
+    }
+
+    // Checking if the pinHandle has called the appropriate method
```
Comment (g-harel) [inline test/unitTest/java/org/thoughtcrime/securesms/pinmessages/PinnedMesssagesHandlerTests.java]: no comments + redundant

**Comment by veken1199** [comment]
I believe it is all good, we have made small discussions regarding almost every single commit

File: test/unitTest/java/org/thoughtcrime/securesms/PinnedMessagesMocks.java (hunk 15 above)
Comment (veken1199) [reply test/unitTest/java/org/thoughtcrime/securesms/PinnedMessagesMocks.java]: @g-harel  yes, it helps to separate the setups from the actual tests

File: src/org/thoughtcrime/securesms/OnSwipeTouchListener.java (hunk 4 above)
Comment (veken1199) [reply src/org/thoughtcrime/securesms/OnSwipeTouchListener.java]: this is a redundant class, we were using to implement a feature differently using swipes @g-harel

File: build.gradle (hunk 2 above)
Comment (veken1199) [reply build.gradle]: I believe this commit is necessary

File: src/org/thoughtcrime/securesms/OnSwipeTouchListener.java (hunk 4 above)
Comment (shelbaz) [reply src/org/thoughtcrime/securesms/OnSwipeTouchListener.java]: redundant, can be removed

File: test/unitTest/java/org/thoughtcrime/securesms/BaseUnitTest.java (hunk 11 above)
Comment (g-harel) [reply test/unitTest/java/org/thoughtcrime/securesms/BaseUnitTest.java:28]: nvm

File: src/org/thoughtcrime/securesms/OnSwipeTouchListener.java (hunk 3 above)
Comment (veken1199) [reply src/org/thoughtcrime/securesms/OnSwipeTouchListener.java]: same, this is redundant class, we might need it in the future tho

File: src/org/thoughtcrime/securesms/PinMessagesLoader.java [hunk 20]
```diff
@@ -0,0 +1,32 @@
+package org.thoughtcrime.securesms;
```
Comment (veken1199) [inline src/org/thoughtcrime/securesms/PinMessagesLoader.java:1]: @g-harel the naming should be PinnedMessageLoader @erinb90 @Dan-Yoo

Comment (g-harel) [reply src/org/thoughtcrime/securesms/PinMessagesLoader.java:1]: you want the package to be different ?

**Review by veken1199 — APPROVED** [review]
I believe some improvements should be committed soon, overall I think it is good

**Review by Radu-Raicea — APPROVED** [review]

File: src/org/thoughtcrime/securesms/PinMessagesLoader.java (hunk 20 above)
Comment (veken1199) [reply src/org/thoughtcrime/securesms/PinMessagesLoader.java:1]: just the class name.

**Pull request merged**